Add support to percolate query to percolate multiple documents simultaneously

martijnvg opened the pull request:
The percolator will add a `_percolator_document_slot` field to all percolator
hits to indicate with what document it has matched. This number matches with
the order in which the documents have been specified in the percolate query.

## Example usage

Create index and index percolator queries:

```
PUT queries
{
  "mappings": {
    "query": {
      "properties": {
        "query" : {
          "type": "percolator"
        },
        "body": {
          "type": "text"
        }
      }
    }
  }
}

PUT /queries/query/1
{
  "query": {
    "match": {
      "body": "brown fox"
    }
  }
}

PUT /queries/query/2
{
  "query": {
    "match": {
      "body": "lazy dog"
    }
  }
}
```

Search request:

```
GET _search
{
  "query": {
    "percolate": {
      "field": "query",
      "documents": [
        {
          "body": "brown dog"
        },
        {
          "body": "pink elephants"
        },
        {
          "body": "lazy fox"  
        }
      ]
    }
  }
}
```

Search response:

```json
{
  "took": 134,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": 0.9808292,
    "hits": [
      {
        "_index": "queries",
        "_type": "query",
        "_id": "2",
        "_score": 0.9808292,
        "_source": {
          "query": {
            "match": {
              "body": "lazy dog"
            }
          }
        },
        "fields": {
          "_percolator_document_slot": [
            0,
            2
          ]
        }
      },
      {
        "_index": "queries",
        "_type": "query",
        "_id": "1",
        "_score": 0.9808292,
        "_source": {
          "query": {
            "match": {
              "body": "brown fox"
            }
          }
        },
        "fields": {
          "_percolator_document_slot": [
            0,
            2
          ]
        }
      }
    ]
  }
}
```

The `_percolator_document_slot` fields indicate that only the first and last documents that have been specified in the percolator query have matched with these two percolator queries. 

The only thing missing here are docs. The percolate docs need to be updated with the fact that now multiple documents can be percolated simultaneously.

jpountz commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -464,7 +525,7 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
         }
 
-        if (document == null) {
+        if (documents == null) {
```
should is rather check for isEmpty?

jpountz commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:
```diff
@@ -0,0 +1,91 @@
+            List<Object> slots = Arrays.stream(topDocs.scoreDocs)
+                .map(scoreDoc -> scoreDoc.doc)
+                .collect(Collectors.toList());
+            fields.put(FIELD_NAME, new DocumentField(FIELD_NAME, slots));
```
I don't think reusing Lucene doc ids as slots will work in case of nested docs.

martijnvg replied:
ah good point. Let me think about this.

martijnvg commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -546,19 +611,26 @@ public String getDocumentType() {
+                .map(ParsedDocument::docs)
+                .flatMap(Collection::stream)
+                .iterator();
+            Iterable<ParseContext.Document> iterable = () -> iterator;
```
Agreed and I will change this. However in this context it will never be invoked multiple times, so it shouldn't cause any issues.

martijnvg commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -413,7 +412,9 @@ public final void testUnknownObjectException() throws IOException {
                     XContentParser.Token token;
                     while ((token = parser.nextToken()) != null) {
-                        if (token == XContentParser.Token.START_OBJECT) {
+                        if (token == XContentParser.Token.START_ARRAY) {
```
@cbuescher Would you mind checking the changes in the file out? I had to change this to make the `testUnknownObjectException()` for the percolator query builder to work.

The percolate query builder now allows multiple documents to percolate via an object array.
Since any content is allowed under this array, this was a suffle protected field. However
these fields were not picked up once this test logic got inside an array. This change should fix that.

cbuescher replied:
@martijnvg Took me a bit to get back into understanding what this weird test does, but now it looks good (at least the changes in this file and the adding of "documents" to the exception list in PercolateQueryBuilderTests). I assume this works well with the other tests AbstractQueryTestCase.

martijnvg replied:
@cbuescher Thanks for checking. Yes, this change does also work with other tests (`gradle core::test` passed).

martijnvg commented:
@jpountz I've addressed your comments, found a way to make mapping lucene docids to slots work when percolating documents with nested objects too and added docs.

jpountz commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -277,8 +299,16 @@ protected void doWriteTo(StreamOutput out) throws IOException {
+                out.writeBytesReference(document);
+            }
+        } else {
+            BytesReference doc = documents.isEmpty() ? null : documents.iterator().next();
```
maybe throw an exception if documents.size() > 1? I believe this could happen in a mixed-version cluster?

jpountz commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java:
```diff
@@ -0,0 +1,116 @@
+        for (; docId > 0; docId = rootDocs.prevSetBit(docId - 1)) {
+            slot++;
+        }
+        return slot;
```
This will need to be repeated for every hit and runs in linear time with the number of percolated documents. Maybe we could decrease the cost of this operation by building an int[] that stores the doc ids of root documents up-front and then using Arrays.binarySearch to find the slot?

martijnvg replied:
Good idea, I'll try that.

martijnvg commented:
> While reasing the changes I wondered what would happen if the top-level query had two sub percolator queries, is it something that we allow today?

@jpountz Yes, that is allowed today. This as things are today can already confuse highlighting (because always the first found percolate query is used). Maybe we shouldn't allow it anymore and fail with an error? (If we chose to do this then I prefer to do this outside this PR, but prior to merging this PR)

martijnvg commented:
@jpountz Do you mean if the user specified multiple percolate queries in a single request? The highligh field name in the response always includes a slot in its name (which will be 0 in case of a single document being percolated (regardless if specified via json object or json array)). Only in the case that multiple percolate queries are specified than the field name changes to indicate which percolate query the highligh hits belong to. Alternatively this can be done in all cases? (single or multiple percolate queries)

jpountz commented:
I'm referring to the fact that the highlight fetch phase only puts the slot in the field name if there are multiple documents to percolate.

martijnvg commented:
Right, in that case too, the highlight field name in the response is different to indicate to which document the highlights belong too. Not sure what do here either, it is choosing between making things more complicated for the case just a single document is percolated and when multiple documents are percolated. I'll update the docs around this specifically to reduce the confusion.
